### PR TITLE
Introduce section on driver polymorphism

### PR DESCRIPTION
**Pull request by AGlass0fMilk**
This commit introduces documentation for the experimental driver polymorphism being added to Mbed OS. See related issue there: ARMmbed/mbed-os#13209

### CODE REVIEW

File: docs/api/drivers/drivers.md [hunk 1]
```diff
@@ -21,3 +21,135 @@ For example, you can attach a callback function on a [Serial](serial.html) objec
...
+When driver polymorphism is enabled, Mbed's driver implementations are marked `final`. This accomplishes two things: 
+
+1. It prevents any further inheritance from the `final` class. 
+2. It hints to the compiler that a pointer or reference to this class will never point to anything but an instance of this class.
+
+The latter effect allows us to minimize the impact of virtual inheritance on execution speed in some cases. If the compiler knows the concrete type of the driver object at runtime, it _may_ eliminate or inline the vtable redirection. In our testing, we have seen this optimization result in _no speed hit_ when toggling a pin on and off in a while loop. Your results may vary depending on your target's HAL implementation, however.
+
+To improve the compiler's ability to optimize calls to a polymorphic driver in this way, you should use the following guidelines when designing your 3rd-party libraries:
+
+When your library **does not** have any high speed requirements, you should use a pointer to the _driver interface_ (eg: `mbed::interface::DigitalOut*`). This gives your library maximum flexibility by allowing it to use polymorphic subtypes of the given interface.
```
Comment (pan-) [inline docs/api/drivers/drivers.md]:
```suggestion
When your library **does not** have any high speed requirements, you should use a pointer/reference to the _driver interface_ (eg: `mbed::interface::DigitalOut*`). This gives your library maximum flexibility by allowing it to use polymorphic subtypes of the given interface.
```

File: docs/api/drivers/drivers.md [hunk 2]
```diff
@@ -21,3 +21,135 @@ For example, you can attach a callback function on a [Serial](serial.html) objec
...
+
+Let's get back to the above example. What you can do now, with driver polymorphism, is make it possible for your external I2C-connected GPIO expander to _act like_ a `DigitalOut` and _be compatible_ with any API that uses `mbed::interface:DigitalOut`:
+
+<span class="images">![](../../images/driver_polymorphism_example.png)
+
+This hierarchy minimizes the changes you need to make to your existing library.
+
+The recommended pattern for designing 3rd-party libraries that take advantage of driver polymorphism is to use _pointers_ to the _interface_ of the driver you're using. For example, we can refactor the `LEDAnimator` class to the following:
+
+```
```
Comment (pan-) [inline docs/api/drivers/drivers.md]:
```suggestion
```c++
```

File: docs/api/drivers/drivers.md [hunk 3]
```diff
@@ -21,3 +21,135 @@ For example, you can attach a callback function on a [Serial](serial.html) objec
...
+
+### Backwards Compatibility
+
+Since driver polymorphism is still experimental, it is disabled by default. Therefore, Mbed's developers have been careful to make sure any code using the existing, non-polymorphic driver classes is not affected by the introduction of this new class hierarchy.
+
+The inheritance of a driver from the new driver interface is conditional upon `FEATURE_EXPERIMENTAL_API` being added to your project's configuration. However, it is still possible for you to develop APIs now that are compatible with both polymorphic and non-polymorphic drivers.
+
+When `FEATURE_EXPERIMENTAL_API` is disabled, the following type alias is made instead:
+
+```
```
Comment (pan-) [inline docs/api/drivers/drivers.md]:
```suggestion
```c++
```

**Review by pan- — APPROVED** [review]
Excellent piece of documentation. It details everything you need to know when polymorphism is enabled.

File: docs/api/drivers/drivers.md (hunk 1 above)
Comment (AGlass0fMilk) [reply docs/api/drivers/drivers.md]: Honestly I had originally written "pointer/reference" but I wasn't sure. To be clear, assigning a subclass to a base class reference does not do any object slicing?

**Comment by AGlass0fMilk** [comment]
What needs to be done to get this merged?

**Pull request merged**